### ISSUE TITLE
What is the correct way to build a mini-batch feed forward network using python?

### ISSUE THREAD

**Issue by easonnie** [clab/dynet#429]
I want to build a mini-batch feed forward network.
```
def train_batch_parallel(batch_state, batch_action, batch_rewards):

        batch_size = len(batch_state)
        input = dy.matInput(input_size, batch_size) # input_size is the dimension of state
        b_state = np.concatenate(batch_state)
        print(b_state.shape)
        input.set(b_state)
        print(input.dim())

        W_exp = dy.parameter(self.W_sm)
        b_exp = dy.parameter(self.b_sm)

        mlp_W_1 = dy.parameter(self.mlp_W_1)
        mlp_b_1 = dy.parameter(self.mlp_b_1)

        mlp_W_2 = dy.parameter(self.mlp_W_2)
        mlp_b_2 = dy.parameter(self.mlp_b_2)

        out_mlp_1 = dy.rectify(dy.affine_transform([mlp_b_1, mlp_W_1, input]))
        out_mlp_2 = dy.rectify(dy.affine_transform([mlp_b_2, mlp_W_2, out_mlp_1]))

        pred = dy.softmax(dy.affine_transform([b_exp, W_exp, out_mlp_2]))
        print(pred.dim())
        print(dy.pick_batch(pred, batch_action).npvalue())
```
I got a `Assertion failed: (pvals->size() == fx.d.batch_elems()), function forward_dev_impl, file /Users/Eason/dynet-base/dynet/dynet/nodes.cc, line 1397.` when run script below.
```
np.random.seed(8)
state1 = np.random.randn(300)
state2 = np.random.randn(300)
train_batch_parallel(batch_state=[state1, state2], batch_action=[1, 0], batch_rewards=[])
```
The result of the code is below and I know the assertion failure was caused by `pick_batch`.
```
[dynet] random seed: 6
[dynet] allocating memory: 4096MB
[dynet] memory allocation done.
Assertion failed: (pvals->size() == fx.d.batch_elems()), function forward_dev_impl, file /Users/Eason/dynet-base/dynet/dynet/nodes.cc, line 1397.
b_state_shape: (600,)
input_dim: (600, 300, 2, 1)
W_exp_dim: (1200, 4, 300, 1)
pred_dim: (8, 4, 2, 1)
```
The correct dimension:
state - {300}
input - {300, 2} with batch
mlp_W_1 - {300, 300}
mlp_b_1 - {300}
mlp_W_2 - {300, 300}
mlp_b_2 - {300}
W_exp - {300, 4}
b_exp - {4}
pred - {4, 2} with batch
out_1 = relu ( mlp_W_1  * input + mlp_b_1 )
out_2 = relu ( mlp_W_2  * out_1 + mlp_b_2 )
pred = softmax ( W_exp * out_2 + b_exp )
loss = - reward * log ( pick(pred, action) ) # policy gradient

What is the correct way to build such model? Any suggestion?

**Comment by pmichel31415** [comment]
It seems like you're trying to set the batch size as a dimension of the input? In Dynet the batch size is handled separately. (Basically it is the "1" you can see at the end of each dim tuple here)

Anyway matinput is deprecated in favor of inputTensor

Here you can probably replace 

    batch_size = len(batch_state)
    input = dy.matInput(input_size, batch_size) # input_size is the dimension of state
    b_state = np.concatenate(batch_state)
    input.set(b_state)

With

    input=dy.inputTensor(batch_state, batched=True)

**Comment by easonnie** [comment]
@pmichel31415 
Thank you so much.
It seems that the documentation did not have such APIs. Is there any example from which I can learn those APIs?

**Comment by pmichel31415** [comment]
Yes we are currently looking for a way to host the python documentation #169 

Until then you can use the `help()` function in the python interpreter

**Comment by easonnie** [comment]
Thanks!
